### PR TITLE
Add detailed_guide_sync_check

### PR DESCRIPTION
**Pull request by tvararu**
Part of migrating detailed guides.

Do not merge as I have not yet run this on the whole dataset; the `content-store` lookups are taking an inordinate amount of time (about 4.5 seconds per check).

I moved the `related_mainstream` and `national_applicability` functions to the model layer because otherwise I'd have to instantiate the presenter as part of the sync checks to verify those two bits. No other sync checks have to do this so I assumed it's undesirable.

The `related_guides` test will currently fail on certain (2012-era mostly) items. These items have references to `related_guides` whole final state is `superseded` or another unpresentable state. The `publishing-api` filters their `content_id`s away, which causes them to not show up in the final `content_store_payload`. This fails the sync check, but there is no clean way to fix it currently, as Whitehall doesn't implement that business logic and can't easily make use of it.

@boffbowsh has expressed interest in fixing this properly across other models that have the same issue (such as HtmlPublications) by fixing the erroneous references.

### CODE REVIEW

File: lib/data_hygiene/publishing_api_sync_check.rb [hunk 1]
```diff
@@ -76,7 +76,7 @@ def override_base_path(&base_path_builder)
     def perform(output: true)
       scope.find_each do |whitehall_model|
         queue_check("content-store", whitehall_model, output)
-        queue_check("draft-content-store", whitehall_model, output)
+        queue_check("draft-content-store", whitehall_model, output) if whitehall_model.draft?
```
Comment (tvararu) [inline lib/data_hygiene/publishing_api_sync_check.rb]: Not sure about this, but I kept getting models that aren't drafts being compared to published versions from the `draft-content-store` which have things like updated titles and what-not.

Comment (boffbowsh) [inline lib/data_hygiene/publishing_api_sync_check.rb]: This change should have fixed this: https://github.com/alphagov/publishing-api/pull/333

Comment (tvararu) [inline lib/data_hygiene/publishing_api_sync_check.rb]: Neat.

**Comment by tvararu** [comment]
After running the checks on the whole dataset, it would seem that a lot of items are failing the `national_applicability` check. Maintaining [Do not merge] until I figure it out.

**Comment by tvararu** [comment]
The failures are largely `related_guides` and `related_mainstream` checks which I assume are failing because of the aforementioned discrepancies between the related content IDs in `whitehall` and the actual available items in the `content-store`.

```
Check complete
Successes: 3684
Failures: 128
```

We can revisit these later.

File: lib/data_hygiene/publishing_api_sync_check.rb [hunk 2]
```diff
@@ -102,8 +102,11 @@ def perform(output: true)
       count = 0
       scope.find_each do |whitehall_model|
         count += 2
-        queue_check("content-store", whitehall_model, output)
-        queue_check("draft-content-store", whitehall_model, output)
+        if whitehall_model.try(:state) == 'draft'
```
Comment (boffbowsh) [inline lib/data_hygiene/publishing_api_sync_check.rb]: I think this should be something like this pseudocode:
```
if model has workflow:
  check the correct content store for this edition
else:
  check both content stores
```

File: script/publishing-api-sync-checks/detailed_guide_sync_check.rb [hunk 3]
```diff
@@ -0,0 +1,78 @@
+detailed_guides = DetailedGuide.latest_published_edition
```
Comment (boffbowsh) [inline script/publishing-api-sync-checks/detailed_guide_sync_check.rb]: This won't check any drafts in the draft content store

File: script/publishing-api-sync-checks/detailed_guide_sync_check.rb [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+check.add_expectation("format") do |content_store_payload, _|
+  content_store_payload["format"] == "detailed_guide"
+end
+
+check.add_expectation("base_path") do |content_store_payload, record|
+  content_store_payload["base_path"] == record.search_link
+end
+
+check.add_expectation("title") do |content_store_payload, record|
+  content_store_payload["title"].strip == record.title.strip
```
Comment (boffbowsh) [inline script/publishing-api-sync-checks/detailed_guide_sync_check.rb]: I think we should only strip the Whitehall model's title, this means we test that the title is being stripped correctly before being sent to publishing api.

File: script/publishing-api-sync-checks/detailed_guide_sync_check.rb [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+end
+
+check.add_expectation("related_guides") do |content_store_payload, record|
+  payload_ids = content_store_payload["links"]["related_guides"]
```
Comment (boffbowsh) [inline script/publishing-api-sync-checks/detailed_guide_sync_check.rb]: RUBY GOLF ALERT
```ruby
content_store_payload["links"]["related_guides"].map { |rg| rg["content_id"] }.to_set == record.related_detailed_guide_content_ids.to_set
```

File: script/publishing-api-sync-checks/detailed_guide_sync_check.rb [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+  records_ids = record.related_detailed_guide_content_ids
+    .uniq
+    .sort
+  payload_ids == records_ids
+end
+
+check.add_expectation("related_mainstream") do |content_store_payload, record|
+  record_rm_content_ids = record.related_mainstream
+
+  payload_related_mainstream = content_store_payload["links"]["related_mainstream"]
```
Comment (boffbowsh) [inline script/publishing-api-sync-checks/detailed_guide_sync_check.rb]: Same golfing as above

File: script/publishing-api-sync-checks/detailed_guide_sync_check.rb (hunk 5 above)
Comment (tvararu) [inline script/publishing-api-sync-checks/detailed_guide_sync_check.rb]: I don't think that's golfish, `to_set` is what I would have used had I known it existed!

I'm going to do something like:

```
payload_ids = content_store_payload["links"]["related_guides"].map { |rg| rg["content_id"] }.to_set
payload_ids == record.related_detailed_guide_content_ids.to_set
```

Which is terse and also keeps line length down.

File: script/publishing-api-sync-checks/detailed_guide_sync_check.rb (hunk 3 above)
Comment (tvararu) [inline script/publishing-api-sync-checks/detailed_guide_sync_check.rb]: What scoping should I use instead? This is what `case_study_sync_check` does.

Comment (boffbowsh) [inline script/publishing-api-sync-checks/detailed_guide_sync_check.rb]: `where(state: %w{published draft withdrawn})`

**Comment by boffbowsh** [comment]
:+1:

**Pull request merged**